torch/_inductor/codegen/triton.py

```diff
@@ -2364,63 +2368,103 @@ def scan(
             masks.append(self._load_mask)
         reduction_range_prefix = self.range_trees[-1].prefix
 
-        value = self.cse.generate(
-            self.compute, f"tl.broadcast_to({value}, {self.dense_size_str()})"
-        )
+        masked_values = []
+        broadcasted_values = []
+        accumulators = []
 
-        default = triton_constant(init)
+        combine_helper_fn = self._lift_helper(combine_fn, len(values))
         dim = self.triton_tensor_ndim() - 1
-        acc_type = triton_acc_type(dtype)
-        cond = " & ".join(masks)
 
-        combine_helper_fn = self._lift_helper(combine_fn, 2)
+        for value, init, dtype in zip(values, inits, dtypes):
+            default = triton_constant(init)
+            acc_type = triton_acc_type(dtype)
+            cond = " & ".join(masks)
 
-        def where_cond(value):
-            if not cond:
-                return value
+            def where_cond(value):
+                if not cond:
+                    return value
+                default_tensor = self.cse.generate(
+                    self.body,
+                    f"tl.full({[1] * self.triton_tensor_ndim()}, {default}, {triton_compute_type(dtype)})",
+                )
+                return self.cse.generate(
+                    self.compute, f"tl.where({cond}, {value}, {default_tensor})"
+                )
+
+            value = self.cse.generate(
+                self.compute,
+                f"tl.broadcast_to({value}.to({triton_compute_type(dtype)}), {self.dense_size_str()})",
```

Comment:
NIT: Better to keep this as two `cse.generate` calls in case `value.to(dtype)` can be CSE'd.